### DISCUSSION TITLE
Generation of analytical (symbolic) model

### DISCUSSION THREAD

**Discussion by totosolat** [stack-of-tasks/pinocchio#1506] [Q&A] [answered]
I am not sure i understand correctly the features of pinocchio:

I used [SYMORO](https://github.com/symoro/symoro) in the past in order to obtain analytical models of robots, I would like to do the same thing with pinocchio.

At the moment, I had a look at the various examples but I could only obtain numerical results (with forward-dynamics-derivatives.py for instance). From what I understand, CppADCodeGen generates a standalone "black box" code instead of the symbolic expression I would like to obtain.

1) Is pinocchio able to produce the symbolic models I am trying to obtain, like SYMORO did? For instance, I would like to generate an analytical dynamic model in which appears all the symbolic parameters and variables (joints position, speed, acceleration, links length, mass, inertia, etc.)

2) If it is possible, could someone give some help or examples of what I am trying to do? Am I the only one interested in this usage?

3) Since SYMORO presents a lot of drawbacks, including the fact that it is not well supported anymore, I think it would be interesting to create a "quick start" guide since it does not seem to be straightforward at the moment. I can contribute on this by creating an example for instance.

Thank you

**Comment by jcarpent** [comment, marked as answer]
> 1. Is pinocchio able to produce the symbolic models I am trying to obtain, like SYMORO did? For instance, I would like to generate an analytical dynamic model in which appears all the symbolic parameters and variables (joints position, speed, acceleration, links length, mass, inertia, etc.)

You can use either CppAD or Casadi support for that. To make the variables visible with naming like `mass1` for the mass associated to link1, then you should set `model.inertias[1].mass() = casadi::SX::sym("mass1",1)`. The same thing can be performed for CppAD. Then, when requiring the output of an algo, like ABA, you will see in the computational graph the name `mass1` appearing.
I cannot provide too much guidance on this task at this time, but it is sure feasible with Pinocchio. I let you explore by yourself.

**Comment by AleksandrKashirin** [comment]
Two years pass, still a guide on the obtaining equations of motion is not there. :(

**Reply by jcarpent** [reply]
What do you mean?

**Reply by AleksandrKashirin** [reply]
I mean I am looking as well for a particular example on how to get a set of symbolic differential equations of the system to use in another library, for instance, in do-mpc. Greatly appreciate if you could help me with some advice on here.
Thank you,
Aleksandr
